### PR TITLE
Copr: add add_datetime_and_duration, add_datetime_and_string and add_time_datetime_null

### PR DESCRIPTION
**Pull request by koushiro**
## What have you changed? (mandatory)

Implement builtin function:  `add_datetime_and_duration`, `add_datetime_and_string` and `add_time_datetime_null`.

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

unit test.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

no

## Does this PR affect tidb-ansible update? (mandatory)

no

## Refer to a related PR or issue link (optional)

[#3275](https://github.com/tikv/tikv/issues/3275)

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 1]
```diff
@@ -238,6 +238,69 @@ impl ScalarFunc {
...
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let add = t
+            .get_time()
+            .checked_add_signed(Duration::nanoseconds(d.to_nanos()));
+        if add.is_none() {
+            return Err(box_err!("parse from duration {} overflows", d));
+        }
+        let mut res = t.to_mut().clone();
+        res.set_time(add.unwrap());
```
Comment (ice1000) [inline src/coprocessor/dag/expr/builtin_time.rs]: This `unwrap` call can be avoided by changing the previous `if add.is_none()` to a pattern matching.

Comment (koushiro) [reply src/coprocessor/dag/expr/builtin_time.rs]: @ice1000 Done.

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 2]
```diff
@@ -238,6 +238,71 @@ impl ScalarFunc {
...
+        );
+        let add = match t
+            .get_time()
+            .checked_add_signed(Duration::nanoseconds(d.to_nanos()))
+        {
+            Some(result) => result,
+            None => return Err(box_err!("parse from duration {} overflows", d)),
+        };
+        let mut res = t.to_mut().clone();
+        res.set_time(add);
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_time.rs]: Can we just `t.set_time(add)`?

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 3]
```diff
@@ -238,6 +238,71 @@ impl ScalarFunc {
...
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let mut t: Cow<'a, Time> = try_opt_or!(
+            self.children[0].eval_time(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let s: Cow<'a, [u8]> = try_opt_or!(
+            self.children[1].eval_string(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let d = match MyDuration::parse(s.as_ref(), 6) {
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_time.rs]: Can you explain the constant "6"?

Comment (koushiro) [reply src/coprocessor/dag/expr/builtin_time.rs]: I will replace constant "6" with mysql::MAX_FSP.

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 2 above)
Comment (koushiro) [reply src/coprocessor/dag/expr/builtin_time.rs]: error[E0596]: cannot borrow immutable borrowed content as mutable    
   --> src/coprocessor/dag/expr/builtin_time.rs:296:9
    |
    |         t.set_time(add);
    |         ^ cannot borrow as mutable

```rust
pub fn get_time(&self) -> DateTime<Tz> {...}
pub fn set_time(&mut self, time: DateTime<Tz>) {...}
```

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 4]
```diff
@@ -13,11 +13,11 @@
 
 use super::{EvalContext, Result, ScalarFunc};
 use chrono::offset::TimeZone;
-use chrono::Datelike;
+use chrono::{Datelike, Duration};
 use coprocessor::codec::error::Error;
 use coprocessor::codec::mysql::time::extension::DateTimeExtension;
 use coprocessor::codec::mysql::time::weekmode::WeekMode;
-use coprocessor::codec::mysql::{Time, TimeType};
+use coprocessor::codec::mysql::{self, Duration as MyDuration, Time, TimeType};
```
[the remark this answers is not in the document]
Comment (huachaohuang) [reply src/coprocessor/dag/expr/builtin_time.rs]: All right, then I think we should keep it consistent.

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 2 above)
Comment (huachaohuang) [reply src/coprocessor/dag/expr/builtin_time.rs]: Oh, I see, that is a `Cow::Borrowed`, never mind.

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 3 above)
Comment (huachaohuang) [reply src/coprocessor/dag/expr/builtin_time.rs]: I see that TiDB use `types.GetFsp()` here, so I think we should use `Time::parse_fsp()` in TiKV instead?

Comment (koushiro) [reply src/coprocessor/dag/expr/builtin_time.rs]: Done.

**Comment by koushiro** [comment]
PTAL again. @ice1000  @huachaohuang

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 5]
```diff
@@ -238,6 +238,72 @@ impl ScalarFunc {
...
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let mut t: Cow<'a, Time> = try_opt_or!(
+            self.children[0].eval_time(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let cow_s: Cow<'a, [u8]> = try_opt_or!(
+            self.children[1].eval_string(ctx, row),
+            Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz)))
+        );
+        let s = box_try!(String::from_utf8(cow_s.to_vec()));
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_time.rs]: How about `str::from_utf8()`?

Comment (koushiro) [reply src/coprocessor/dag/expr/builtin_time.rs]: Done.

**Comment by huachaohuang** [comment]
I think the sign-off is missing in some of your commits, you can squash your commits into one and add a sign-off, and then force push it.

**Comment by koushiro** [comment]
> I think the sign-off is missing in some of your commits, you can squash your commits into one and add a sign-off, and then force push it.

Sign-off has been added. Thanks.

**Comment by huachaohuang** [comment]
@ice1000 PTAL

**Comment by koushiro** [comment]
@ice1000 @huachaohuang  conflict was resolved, PTAL again.

**Comment by huachaohuang** [comment]
@ice1000 PTAL

**Pull request merged**